---
title: "bcb546x_r_assignments"
author: "Zihao Zheng"
date: "10/9/2018"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
## Part I
### 1. Data inspection
```{r message=FALSE, warning=FALSE}
setwd("~/hmwk/")
library(tidyverse)
library(reshape2)
library(ggrepel)
```

Check the size of files to be loaded
```{r}

file.size("fang_et_al_genotypes.txt") %>% utils:::format.object_size("auto")
file.size("snp_position.txt") %>% utils:::format.object_size("auto")

```


load files to R
```{r message=FALSE}

fang_et_al <- read_delim("fang_et_al_genotypes.txt",delim = "\t",col_names = T)
snp_position <- read_delim("snp_position.txt",delim = "\t",col_names = T)

```

number of columns and rows for each file

```{r}
# fang_et_al column#
ncol(fang_et_al)
# fang_et_al row#
nrow(fang_et_al)

head(fang_et_al)

# snp_position column#
ncol(snp_position)
# snp_position row#
nrow(snp_position)

head(snp_position)

```

### 2. Data processing

split `fang_et_al` into maize group and teosinte group
```{r}

maize.snp <- fang_et_al %>% filter(Group %in% c("ZMMIL","ZMMLR","ZMMMR"))
teosinte.snp <- fang_et_al %>% filter(Group %in% c("ZMPBA","ZMPIL","ZMPJA"))

```

transpose genotypic data, and merge with the `snp_position`
```{r message=FALSE}
# maize group

maize.tmp <- data.frame(t(maize.snp)) %>% tibble::rownames_to_column()
colnames(maize.tmp) <- maize.tmp[1, ] 
maize.tmp <- maize.tmp[-(1:3),] 
colnames(maize.tmp)[1] <- "SNP_ID"
merged.maize.genotype <- merge(snp_position,maize.tmp,by = "SNP_ID")

# teosinte group
teosinte.tmp <- data.frame(t(teosinte.snp)) %>% tibble::rownames_to_column()
colnames(teosinte.tmp) <- teosinte.tmp[1, ] 
teosinte.tmp <- teosinte.tmp[-(1:3),] 
colnames(teosinte.tmp)[1] <- "SNP_ID"
merged.teosinte.genotype <- merge(snp_position,teosinte.tmp,by = "SNP_ID")

```


For maize group, generate `10 files (1 for each chromosome) with SNPs ordered based on increasing position values and with missing data encoded by this symbol: ?`

```{r warning=FALSE}

new_names <- c("maize_chr1","maize_chr10","maize_chr2","maize_chr3","maize_chr4","maize_chr5",
               "maize_chr6","maize_chr7","maize_chr8","maize_chr9","maize_multiple","maize_unknown")

maize_split_incr <- split(merged.maize.genotype, merged.maize.genotype$Chromosome)

for (i in 1:10) {
  maize_split_incr[[i]]<- maize_split_incr[[i]][order(as.numeric(maize_split_incr[[i]]$Position)),]
  write_delim(maize_split_incr[[i]],paste0(new_names[i],"_incr.txt"),delim = "\t")
 }


```


For maize group, generate `10 files (1 for each chromosome) with SNPs ordered based on decreasing position values and with missing data encoded by this symbol: -`

```{r warning=FALSE}

merged.maize.genotype[merged.maize.genotype== "?/?"]<- "-/-"

maize_split_decr <- split(merged.maize.genotype, merged.maize.genotype$Chromosome)

for (i in 1:10) {
  maize_split_decr[[i]]<- maize_split_decr[[i]][order(as.numeric(maize_split_decr[[i]]$Position)),]
  write_delim(maize_split_decr[[i]],paste0(new_names[i],"_decr.txt"),delim = "\t")
 }


```

For teosinte group, generate `10 files (1 for each chromosome) with SNPs ordered based on increasing position values and with missing data encoded by this symbol: ?`

```{r warning=FALSE}

new_names <- c("teosinte_chr1","teosinte_chr10","teosinte_chr2","teosinte_chr3","teosinte_chr4","teosinte_chr5",
               "teosinte_chr6","teosinte_chr7","teosinte_chr8","teosinte_chr9","teosinte_multiple","teosinte_unknown")

teosinte_split_incr <- split(merged.teosinte.genotype, merged.teosinte.genotype$Chromosome)

for (i in 1:10) {
  teosinte_split_incr[[i]]<- teosinte_split_incr[[i]][order(as.numeric(teosinte_split_incr[[i]]$Position)),]
  write_delim(teosinte_split_incr[[i]],paste0(new_names[i],"_incr.txt"),delim = "\t")
 }


```


For teosinte group, generate `10 files (1 for each chromosome) with SNPs ordered based on decreasing position values and with missing data encoded by this symbol: -`

```{r warning=FALSE}

merged.teosinte.genotype[merged.teosinte.genotype== "?/?"]<- "-/-"

teosinte_split_decr <- split(merged.teosinte.genotype, merged.teosinte.genotype$Chromosome)

for (i in 1:10) {
  teosinte_split_decr[[i]]<- teosinte_split_decr[[i]][order(as.numeric(teosinte_split_decr[[i]]$Position)),]
  write_delim(teosinte_split_decr[[i]],paste0(new_names[i],"_decr.txt"),delim = "\t")
 }


```

## Part II

Merge `fang_et_al` and `snp_position` to get a master file
```{r}

genotype.tmp <- data.frame(t(fang_et_al)) %>% tibble::rownames_to_column()
colnames(genotype.tmp) <- genotype.tmp[1, ] 
genotype.tmp <- genotype.tmp[-(1:3),] 
colnames(genotype.tmp)[1] <- "SNP_ID"
merged.genotype <- merge(snp_position,genotype.tmp,by = "SNP_ID")


```

### 1(a). Plot the total number of SNPs in our dataset on each chromosome.
```{r}
p1 <- ggplot(data = merged.genotype) + geom_bar(mapping = aes(x=Chromosome)) + scale_x_discrete(limits=c(1:10,"multiple","unknown")) + theme(text = element_text(size=10))
p1

```


### 1(b).What groups contribute most of these SNPs?
```{r}
fang_et_al_short <- fang_et_al %>% select(-c("Sample_ID","JG_OTU")) %>% melt(id.vars = "Group",variable.name = "SNP_ID") %>% unique()

snp_position.tmp <- snp_position %>% select(SNP_ID,Chromosome) %>% merge(fang_et_al_short,by = "SNP_ID")
snp_position.tmp[snp_position.tmp=="?/?"] <- NA
snp_position.tmp <- na.omit(snp_position.tmp)

snp_group_stat <- snp_position.tmp %>% count(Group)
snp_group_stat$Label <- paste(snp_group_stat$Group,paste(round(((snp_group_stat$n/sum(snp_group_stat$n))*100),2),"%"), sep="-")
snp_group_stat$pos = (cumsum(c(0, snp_group_stat$n)) + c(snp_group_stat$n / 2, .01))[1:nrow(snp_group_stat)]


p2 <- ggplot(snp_group_stat, aes(1, n, fill = Group)) +
    geom_col(color = 'black', 
             position = position_stack(reverse = TRUE), 
             show.legend = TRUE) +
    geom_text_repel(aes(x = 1.4, y = pos, label = Label), 
                    nudge_x = .3, 
                    segment.size = .7, 
                    show.legend = FALSE) +
    coord_polar('y') +
    theme_void()
p2
```

###2. Missing data and amount of heterozygosity
```{r}
fang_et_al_short2 <- fang_et_al %>% select(-one_of("JG_OTU")) %>% melt(id.vars = c("Group","Sample_ID"),variable.name = "SNP_ID")
fang_et_al_short2[fang_et_al_short2=="?/?"] <- NA

fang_et_al_short2$allele[is.na(fang_et_al_short2$value)] <- NA
fang_et_al_short2$allele[fang_et_al_short2$value %in% c("A/A","C/C","G/G","T/T")] <- "homozygous"
fang_et_al_short2$allele[!fang_et_al_short2$value %in% c("A/A","C/C","G/G","T/T",NA)] <- "heterozygous"


allele_stat_group <- fang_et_al_short2 %>% group_by(Group,allele) %>% 
  summarise(n = n()) %>% mutate(countT= sum(n)) %>% 
  mutate(percentage=round(n/countT,2))

p3 <- ggplot(allele_stat_group) + geom_bar(aes(y = percentage, x = Group, fill = allele),stat="identity")+ theme(axis.text.x=element_text(angle = -90, hjust = 0)) 


allele_stat_sample <- fang_et_al_short2 %>% group_by(Sample_ID,allele) %>% 
  summarise(n = n()) %>% mutate(countT= sum(n)) %>% 
  mutate(percentage=round(n/countT,2))

p4 <- ggplot(allele_stat_sample) + geom_bar(aes(y = percentage, x = Sample_ID, fill = allele),
                           stat="identity")+ 
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
```

### 2(a).Missing data and amount of heterozygosity by group
```{r}
p3
```
### 2(b).Missing data and amount of heterozygosity by individual
```{r}
p4
```


###3.Visualization of Minor Allele Frequency (MAF) of maize group and teosinte group
Calculate MAF for each group
```{r message=FALSE}
MAF.tmp <- fang_et_al_short2 %>% filter(allele=="homozygous") %>% 
  group_by(Group,SNP_ID,value) %>%
  summarise(n = n()) %>% mutate(countT= sum(n)) %>% 
  mutate(allele_freq=round(n/countT,3)) %>% 
  filter(allele_freq < 0.5) %>% 
  select(Group,SNP_ID,allele_freq) %>% 
  group_by(Group,SNP_ID) %>% top_n(-1)

colnames(MAF.tmp)[3] <- "MAF"

MAF.tmp$species <- NA
MAF.tmp$species[MAF.tmp$Group %in% c("ZMMIL","ZMMLR","ZMMMR")] <- "maize"
MAF.tmp$species[MAF.tmp$Group %in% c("ZMPBA","ZMPIL","ZMPJA")] <- "teosinte"
MAF.tmp <- na.omit(MAF.tmp)

p5<- ggplot(MAF.tmp, aes(x=MAF, fill=Group)) +
    geom_histogram(binwidth=0.01, alpha=.5, position="identity") +
    facet_grid(species ~ .)
p5
```






